File: src/lib/output.ts
import { applyTruncation } from "./truncation.ts";
import { getSettings } from "./config.ts";

const DEFAULT_PAGE_SIZE = 50;

export function pruneEmpty<T>(value: T): T {
  const pruned = pruneEmptyInternal(value);
  return (pruned === undefined ? ({} as T) : (pruned as T)) as T;
}

function pruneEmptyInternal(value: unknown): unknown {
  if (value === null || value === undefined) {
    return undefined;
  }

  if (typeof value === "string") {
    return value.trim() === "" ? undefined : value;
  }

  if (typeof value === "number" || typeof value === "boolean") {
    return value;
  }

  if (Array.isArray(value)) {
    const next = value
      .map((v) => pruneEmptyInternal(v))
      .filter((v): v is Exclude<unknown, undefined> => v !== undefined);
    return next.length === 0 ? undefined : next;
  }

  if (typeof value === "object") {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      const next = pruneEmptyInternal(v);
      if (next !== undefined) {
        out[k] = next;
      }
    }
    return Object.keys(out).length === 0 ? undefined : out;
  }

  return value;
}

export function printJson(data: unknown): void {
  console.log(JSON.stringify(applyTruncation(pruneEmpty(data)), null, 2));
}

/**
 * Print paginated list output with { items, pagination? } wrapper.
 * Always returns { items: [...] } even when the array is empty.
 * Includes pagination when hasNextPage is true.
 */
export function printPaginated(
  items: unknown[],
  pageInfo: { hasNextPage: boolean; endCursor?: string },
): void {
  // Prune individual items, apply truncation, but always preserve the array structure
  const prunedItems = items.map((item) => applyTruncation(pruneEmpty(item)));
  const payload: Record<string, unknown> = { items: prunedItems };
  if (pageInfo.hasNextPage) {
    payload.pagination = {
      hasMore: true,
      nextCursor: pageInfo.endCursor ?? null,
    };
  }
  console.log(JSON.stringify(payload, null, 2));
}

export function printError(message: string): void {
  console.error(JSON.stringify({ error: message }));
  process.exitCode = 1;
}

export function resolvePageSize(opts: { limit?: string }): number {
  if (opts.limit !== undefined) {
    return parseInt(opts.limit, 10);
  }
  const settings = getSettings();
  return settings.pagination?.defaultPageSize ?? DEFAULT_PAGE_SIZE;
}

File: internal/output/output.go
package output

import (
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/spf13/cobra"

	"github.com/shhac/lin/internal/config"
	"github.com/shhac/lin/internal/truncation"
)

const DefaultPageSize = 50

type Pagination struct {
	HasMore    bool   `json:"hasMore"`
	NextCursor string `json:"nextCursor,omitempty"`
}

type PaginatedResult struct {
	Items      any         `json:"items"`
	Pagination *Pagination `json:"pagination,omitempty"`
}

func PrintJSON(data any) {
	b, err := json.Marshal(data)
	if err != nil {
		return
	}
	var decoded any
	if err := json.Unmarshal(b, &decoded); err != nil {
		return
	}
	decoded = truncation.Apply(pruneEmpty(decoded))
	enc := json.NewEncoder(os.Stdout)
	enc.SetIndent("", "  ")
	enc.SetEscapeHTML(false)
	_ = enc.Encode(decoded)
}

func PrintPaginated(items any, pageInfo *Pagination) {
	b, err := json.Marshal(items)
	if err != nil {
		return
	}
	var decoded any
	if err := json.Unmarshal(b, &decoded); err != nil {
		return
	}
	// Prune and truncate each item individually
	if arr, ok := decoded.([]any); ok {
		for i, item := range arr {
			arr[i] = truncation.Apply(pruneEmpty(item))
		}
		decoded = arr
	}
	result := map[string]any{"items": decoded}
	if pageInfo != nil && pageInfo.HasMore {
		result["pagination"] = map[string]any{
			"hasMore":    true,
			"nextCursor": pageInfo.NextCursor,
		}
	}
	enc := json.NewEncoder(os.Stdout)
	enc.SetIndent("", "  ")
	enc.SetEscapeHTML(false)
	_ = enc.Encode(result)
}

func PrintError(msg string) {
	enc := json.NewEncoder(os.Stderr)
	enc.SetEscapeHTML(false)
	_ = enc.Encode(map[string]string{"error": msg})
	os.Exit(1)
}

func PrintErrorf(format string, args ...any) {
	PrintError(fmt.Sprintf(format, args...))
}

// ResolveCursor returns a *string for pagination, or nil if empty.
func ResolveCursor(cursor string) *string {
	if cursor == "" {
		return nil
	}
	return &cursor
}

// ResolvePageSize returns the page size from --limit flag, config, or default.
func ResolvePageSize(limit string) int {
	if limit != "" {
		n, err := strconv.Atoi(limit)
		if err == nil && n > 0 {
			return n
		}
	}
	cfg := config.Read()
	if cfg.Settings != nil && cfg.Settings.Pagination != nil && cfg.Settings.Pagination.DefaultPageSize != nil {
		return *cfg.Settings.Pagination.DefaultPageSize
	}
	return DefaultPageSize
}

// HandleUnknownCommand registers a handler for unknown subcommands on a cobra command.
func HandleUnknownCommand(cmd *cobra.Command, hint string) {
	cmd.RunE = func(cmd *cobra.Command, args []string) error {
		if len(args) > 0 {
			var names []string
			for _, sub := range cmd.Commands() {
				if sub.Name() != "usage" && sub.Name() != "help" {
					names = append(names, sub.Name())
				}
			}
			msg := fmt.Sprintf("Unknown command: %q. Valid commands: %s", args[0], strings.Join(names, ", "))
			if hint != "" {
				msg += ". " + hint
			}
			PrintError(msg)
		}
		return cmd.Help()
	}
}

func pruneEmpty(v any) any {
	if v == nil {
		return nil
	}
	switch val := v.(type) {
	case map[string]any:
		out := make(map[string]any, len(val))
		for k, v := range val {
			if v == nil {
				continue
			}
			pruned := pruneEmpty(v)
			if isEmpty(pruned) {
				continue
			}
			out[k] = pruned
		}
		if len(out) == 0 {
			return nil
		}
		return out
	case []any:
		out := make([]any, 0, len(val))
		for _, v := range val {
			pruned := pruneEmpty(v)
			if pruned != nil {
				out = append(out, pruned)
			}
		}
		return out
	case string:
		if strings.TrimSpace(val) == "" {
			return nil
		}
		return val
	default:
		return v
	}
}

func isEmpty(v any) bool {
	switch val := v.(type) {
	case nil:
		return true
	case string:
		return strings.TrimSpace(val) == ""
	case map[string]any:
		return len(val) == 0
	case []any:
		return len(val) == 0
	default:
		return false
	}
}
